src/validation.h

```diff
@@ -893,6 +893,8 @@ class ChainstateManager
     //! that call.
     std::unique_ptr<Chainstate> m_snapshot_chainstate GUARDED_BY(::cs_main);
 
+    std::array<ThresholdConditionCache, VERSIONBITS_NUM_BITS> m_warningcache GUARDED_BY(::cs_main);
```

Comment:
It might be better if this were `private`. I guess something that would be possible if you moved the `for (int bit = 0; bit < VERSIONBITS_NUM_BITS; bit++) { ... }` loop into a method like:

```c++
std::vector<std::tuple<ThresholdState, int>> CheckWarnings(const CBlockIndex* pindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```

then looped through the resulting vector to pull out `state` and `bit`.